pkg/reconciler/common/releases.go

```diff
@@ -0,0 +1,140 @@
+/*
+Copyright 2020 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package common
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path"
+	"path/filepath"
+	"sort"
+
+	mf "github.com/manifestival/manifestival"
+	"golang.org/x/mod/semver"
+	"knative.dev/operator/pkg/apis/operator/v1alpha1"
+)
+
+const (
+	KoEnvKey = "KO_DATA_PATH"
+)
+
+var cache = map[string]mf.Manifest{}
+
+func TargetVersion(instance v1alpha1.KComponent) string {
+	switch {
+	case instance.GetSpec().GetVersion() != "":
+		return instance.GetSpec().GetVersion()
+	case instance.GetStatus().GetVersion() != "":
```

Comment:
@jcrossley3 @markusthoemmes 
Again, I personally lean on to this approach. If spec.version is nil, go to check status.version. 
We discussed multiple time in slack channel. It seemed both of you are for directly looking up latestRelease, if spec.version is nil.

Reply:
I agree we need to deal with this, but I was thinking this PR wouldn't care since `spec.Version` isn't yet in the CRD. Until it is, both `TargetManifest` and `InstalledManifest` will return the `latestRelease`. But that will still be a problem for any existing CR's with a `status.Version` unknown to this operator. I'm not sure how `FinalizeKind` handles errors, but I'd expect that situation to put us in an infinite loop. Maybe if `InstalledManifest` returns an error, we use `TargetManifest` instead and log a warning that some cluster-scoped resources might not get cleaned up?

Reply:
We can return a warning message for now.

Reply:
I've added the above logic to `FinalizeKind` in both serving and eventing -- there's opportunity to factor them into a single common function, but I want to get some consensus on the correctness of the logic for the purposes of this PR. Thoughts?

Reply:
This PR added the cache with the path as key, the logic to retrieve the installed and target manifests, added spec.version into the objects...
It looks in the correct direction to me. Further refactoring can be done based on this PR.

Reply:
I agree, this is a change of behavior and not what we've discussed extensively I think. We wanted to always return `latestRelease` if `spec.Version` isn't set.

Reply:
I can do that. I had it in my head that returning `latestRelease` would be a surprise to the user in light of the operator having the multi-version behavior.

Reply:
Done